runtime/lua/vim/_defaults.lua

```diff
@@ -32,27 +32,59 @@ do
   ---
   --- See |v_star-default| and |v_#-default|
   do
-    local function _visual_search(cmd)
-      assert(cmd == '/' or cmd == '?')
-      local chunks =
-        vim.fn.getregion(vim.fn.getpos('.'), vim.fn.getpos('v'), { type = vim.fn.mode() })
+    local function _visual_search(forward)
+      assert(forward == 0 or forward == 1)
+      local pos = vim.fn.getpos('.')
+      local vpos = vim.fn.getpos('v')
+      local mode = vim.fn.mode()
+      local chunks = vim.fn.getregion(pos, vpos, { type = mode })
       local esc_chunks = vim
         .iter(chunks)
         :map(function(v)
-          return vim.fn.escape(v, cmd == '/' and [[/\]] or [[?\]])
+          return vim.fn.escape(v, [[\]])
         end)
         :totable()
       local esc_pat = table.concat(esc_chunks, [[\n]])
-      local search_cmd = ([[%s\V%s%s]]):format(cmd, esc_pat, '\n')
-      return '\27' .. search_cmd
+      if #esc_pat == 0 then
+        vim.api.nvim_echo({ { 'E348: No string under cursor' } }, true, { err = true })
+        return '<Esc>'
+      end
+      local search = [[\V]] .. esc_pat
+
+      vim.fn.setreg('/', search)
+      vim.fn.histadd('/', search)
+      vim.v.searchforward = forward
+
+      -- The count has to be adjusted when searching backwards and the cursor
+      -- isn't positioned at the beginning of the selection
+      local count = vim.v.count1
+      if
+        forward == 0
+        and (function()
+          local _, line, col, _ = unpack(pos)
+          local _, vline, vcol, _ = unpack(vpos)
+          if line > vline then
+            return true
+          elseif mode == 'v' then
+            return line == vline and col > vcol
+          elseif mode == 'V' then
+            return col ~= 1
+          elseif mode == '\22' then
+            return col > vcol
+          end
+        end)()
+      then
+        count = count + 1
+      end
+      return '<Esc>' .. count .. 'n'
     end
 
     vim.keymap.set('x', '*', function()
```

Comment:
Not sure if this is intended, but the usual * and # ones work on a word boundaries, adding < and > to the pattern, and the g* and g# variations do not, do we need to support visual g* and g# ?

Reply:
I think it makes sense to not use word boundaries since visual `*`/`#` is much more deliberate (you see the boundaries of the selection before you search) than normal mode `*`/`#`.